src/logql-query.ts

```diff
@@ -1,48 +1,250 @@
+          if (node.name === 'Selector' && parsedMatchers.length === 0) {
+            parsedMatchers = parseMatchers(syntaxTree, node, query);
+
+            return false;
+          } else if (
+            node.name === 'PipelineExpr' &&
+            parsedPipeline.length === 0
+          ) {
+            parsedPipeline = parsePipelineStages(syntaxTree, node, query);
+
+            return false;
+          }
+        }
+      },
+    });
+
+    this.streamSelector = parsedMatchers;
+    this.pipeline = parsedPipeline;
   }
 
-  appendSelector = (newSelectors: string | Array<string>) => {
-    if (Array.isArray(newSelectors)) {
-      this.selectors.push(...newSelectors);
-    } else {
-      this.selectors.push(newSelectors);
+  removeSelectorMatcher = (matcherToRemove: Partial<LabelMatcher>) => {
+    this.streamSelector = this.streamSelector.filter((matcher) => {
+      let include = true;
+
+      if (matcherToRemove.label) {
+        include = include && matcher.label !== matcherToRemove.label;
+      }
+
+      return include;
+    });
+
+    return this;
+  };
+
+  addSelectorMatcher = (matchers?: LabelMatcher | Array<LabelMatcher>) => {
+    if (!matchers) {
+      return this;
     }
 
-    this.selectors = this.selectors.filter(notEmptyString).map(trim);
+    const matchersArray = Array.isArray(matchers) ? matchers : [matchers];
+
+    matchersArray.forEach((matcher) => {
+      const existingSelectorIndex = this.streamSelector.findIndex(
+        ({ label }) => matcher.label === label,
+      );
+
+      if (existingSelectorIndex !== -1) {
+        this.streamSelector[existingSelectorIndex] = matcher;
+      } else {
+        this.streamSelector.push(matcher);
+      }
+    });
+
+    this.streamSelector = this.streamSelector.filter(
+      (selector, index) =>
+        this.streamSelector.findIndex(
+          ({ label }) => selector.label === label,
+        ) === index,
+    );
+
+    return this;
+  };
+
+  removePipelineStage = (
+    pipelineStageToRemove: Partial<PipelineStage>,
+    matchOptions?: { matchLabel?: string },
+  ) => {
+    this.pipeline = this.pipeline.filter((existingStage) => {
+      let include = true;
+
+      if (pipelineStageToRemove.operator) {
+        include =
+          include && existingStage.operator !== pipelineStageToRemove.operator;
+      }
+
+      if (pipelineStageToRemove.value) {
+        include =
+          include && existingStage.value !== pipelineStageToRemove.value;
+      }
+
+      if (matchOptions?.matchLabel) {
+        include =
+          include &&
+          !existingStage.labelsInFilter?.every(
+            (label) => label.label === matchOptions.matchLabel,
+          );
+      }
+
+      return include;
+    });
 
     return this;
   };
 
-  appendPipeline = (newPipeline: string | Array<string>) => {
-    if (Array.isArray(newPipeline)) {
-      this.pipeline.push(...newPipeline);
+  addPipelineStage = (
+    newPipelineStage?: PipelineStage | Array<PipelineStage>,
```

Comment:
Similar to `addSelectorMatcher`, should `newPipelineStage` be optional? Calling this function without it wouldn't do anything useful.

Reply:
pipeline stages are created in other functions. These other functions can return undefined when the creation was not successful, for example due to filters being undefined. IMO this signature is more resilient as in fact is filtering out undefined stages.

Reply:
👍